File: sdk/elixir/lib/bsdkrun/volumes.ex
defmodule Bsdkrun.Volumes do
  @moduledoc "Persistent volume operations."

  alias Bsdkrun.{Cli, Error, Types}
  alias Bsdkrun.Types.VolumeInfo

  @doc "List persistent volumes."
  @spec list() :: {:ok, [VolumeInfo.t()]} | {:error, Error.t()}
  def list do
    with {:ok, res} <- Cli.checked(["volume", "ls", "--json"], "bsdkrun volume ls") do
      rows =
        res.stdout
        |> then(&if(String.trim(&1) == "", do: "[]", else: &1))
        |> Jason.decode!()
        |> Enum.map(&Types.volume_info/1)

      {:ok, rows}
    end
  end

  @doc "Remove one or more volumes (and their data)."
  @spec remove(String.t() | [String.t()], keyword()) :: :ok | {:error, Error.t()}
  def remove(names, opts \\ []) do
    list = List.wrap(names)
    base = if Keyword.get(opts, :force, false), do: ["volume", "rm", "--force"], else: ["volume", "rm"]

    case Cli.checked(base ++ list, "bsdkrun volume rm") do
      {:ok, _} -> :ok
      {:error, _} = err -> err
    end
  end
end

File: sdk/python/src/bsdkrun/volumes.py
"""Host-level volume operations."""

from __future__ import annotations

import builtins
import json
from collections.abc import Sequence

from .process import run_checked
from .types import VolumeInfo

__all__ = ["list", "remove"]


def list() -> builtins.list[VolumeInfo]:
    """List persistent volumes."""
    res = run_checked(["volume", "ls", "--json"], "bsdkrun volume ls")
    rows = json.loads(res.stdout or "[]")
    return [VolumeInfo.from_row(row) for row in rows]


def remove(names: str | Sequence[str], *, force: bool = False) -> None:
    """Remove one or more volumes (and their data)."""
    name_list = [names] if isinstance(names, str) else [*names]
    args = ["volume", "rm"]
    if force:
        args.append("--force")
    args += name_list
    run_checked(args, "bsdkrun volume rm")
